src/django_pg_migration_tools/management/commands/migrate_with_timeouts.py

```diff
@@ -28,32 +53,237 @@ def add_arguments(self, parser: base.CommandParser) -> None:
+                "follow this signature: f(retry_state) where `retry_state` "
+                "is the dataclass: "
+                "django_pg_migration_tools.management.commands.migrate_with_timeouts.RetryState."
+                "This callback can be useful for calling loggers after each "
+                "retry."
+            ),
+        )
+        parser.add_argument(
+            "--lock-timeout-max-retries",
+            dest="lock_timeout_max_retries",
+            type=int,
+            required=False,
+            default=0,
+            help=(
+                "How many times to retry after a lock timeout happened. "
+                "Defaults to zero, which means no retries - the migration "
+                "fails immediately upon lock timeout."
+            ),
+        )
+        parser.add_argument(
+            "--lock-timeout-retry-exp",
+            dest="lock_timeout_retry_exp",
+            type=int,
+            required=False,
+            default=2,
+            help="The value for the exponent for retry backoff delay. Defaults to 2.",
+        )
+        parser.add_argument(
+            "--lock-timeout-retry-max-wait-in-ms",
+            dest="lock_timeout_retry_max_wait_in_ms",
+            type=int,
+            required=False,
+            default=10_000,
+            help=(
+                "Sets a limit to the maximum length of time between subsequent "
+                "exponential backoff retries. Defaults to 10s."
+            ),
+        )
+        parser.add_argument(
+            "--lock-timeout-retry-min-wait-in-ms",
+            dest="lock_timeout_retry_min_wait_in_ms",
+            type=int,
+            required=False,
+            default=3_000,
+            help=(
+                "Sets a limit to the minimum length of time between subsequent "
+                "exponential backoff retries. Defaults to 3s."
+            ),
+        )
         super().add_arguments(parser)
 
     @base.no_translations
     def handle(self, *args: Any, **options: Any) -> None:
-        statement_timeout_in_ms: int | None = options["statement_timeout_in_ms"]
-        lock_timeout_in_ms: int | None = options["lock_timeout_in_ms"]
+        timeout_options = MigrationTimeoutOptions.from_dictionary(options)
+        timeout_options.validate()
+        retry_strategy = MigrateRetryStrategy(timeout_options=timeout_options)
 
-        if statement_timeout_in_ms is None and lock_timeout_in_ms is None:
+        migration_applied: bool = False
+        while retry_strategy.can_migrate() and (not migration_applied):
+            try:
+                with timeouts.apply_timeouts(
+                    using=options["database"],
+                    lock_timeout=timeout_options.lock_timeout,
+                    statement_timeout=timeout_options.statement_timeout,
+                ):
+                    super().handle(*args, **options)
+                    migration_applied = True
```

Comment:
```suggestion
                    return
```

And then remove the `migration_applied` flag

Reply:
Good shout db207cdd0ac4b69b1ae4b9570f06191992fd74f5